README.md
# @ox2/thumb
![NPM version](https://img.shields.io/badge/npm-private-orange.svg?style=flat)
<!-- ![NPM version](https://img.shields.io/npm/v/@ox2/thumb.svg?style=flat) -->

Thumbnail react component.

## Installation
Install using [npm](http://npmjs.com):
```sh
npm install @ox2/thumb --save
```
Install using [yarn](http://yarnpkg.com):
```sh
yarn add @ox2/thumb
```

## Usage
```js
import Thumb from '@ox2/thumb/Thumb';

<Thumb height={60} image={image1} width={60} />


```

## Change Log
Changes are tracked in the [CHANGELOG.md](https://github.com/ox2/thumb/blob/master/CHANGELOG.md)

## License
[MIT](https://github.com/ox2/thumb/blob/master/LICENSE)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## Unreleased

## 1.0.0 - 2017-06-27
### Added
- Initial release. 👾👾


src/Thumb/Thumb.test.js
/* eslint-disable no-unused-vars */
import React from 'react';
import { shallow, render, mount } from 'enzyme';
import Thumb from './Thumb';

const image1 =
  'https://randomuser.me/api/portraits/women/50.jpg';
const logo1 =
  'https://upload.wikimedia.org/wikipedia/commons/thumb/c/c6/Rohde_%26_Schwarz_Logo.svg/800px-Rohde_%26_Schwarz_Logo.svg.png';

it('renders correctly as square', () => {
  const wrapper = mount(
    <Thumb height={60} image={image1} width={60} />,
  );

  expect(wrapper).toMatchSnapshot();
});

it('renders correctly as landscape', () => {
  const wrapper = mount(
    <Thumb height={60} image={logo1} width={120} />,
  );

  expect(wrapper).toMatchSnapshot();
});


src/Thumb/Thumb.js
import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';
import FontIcon from '@ox2/icon/FontIcon';
import ImageOffline from '@ox2/image/ImageOffline';

/**
 * Thumb Component
 */
const Thumb = props => {
  const {
    className: classNameProp,
    image,
    icon,
    color,
    iconColor,
    iconGroup,
    width,
    height,
    offlineEnabled,
    localImg,
  } = props;

  const className = classNames(
    'display:flex flex:items-center flex:content-center border:rounded',
    `color:${color}`,
    classNameProp,
  );

  const ICON_SCALE = 0.7;

  return (
    <div
      className={className}
      style={{
        width: `${width}px`,
        height: `${height}px`,
        WebkitFlex: `0 0 ${width}px`,
        flex: `0 0 ${width}px`,
      }}
    >
      {
        do {
          if (image) {
            if (offlineEnabled && localImg) {
              <ImageOffline
                className="w:100p border:rounded"
                localImg={localImg}
                remoteLink={image}
              />;
            } else {
              <img
                className="w:100p border:rounded"
                src={image}
              />;
            }
          } else {
            <FontIcon
              className={`text-color:${iconColor}`}
              group={iconGroup}
              icon={icon}
              size={`${Math.round(height * ICON_SCALE)}px`}
            />;
          }
        }
      }
    </div>
  );
};

Thumb.defaultProps = {
  className: '',
  color: 'transparent',
  iconGroup: '',
  iconColor: 'neutral-light',
  icon: '',
  image: '',
  localImg: '',
  offlineEnabled: false,
};

Thumb.propTypes = {
  /**
   * The css class name of the root element.
   */
  className: PropTypes.string,
  /**
   * Background color
   */
  color: PropTypes.string,
  /**
   * Thumb height in pixels
   */
  height: PropTypes.number.isRequired,
  /**
   * Icon id
   */
  icon: PropTypes.string,
  /**
   * Icon color
   */
  iconColor: PropTypes.string,
  /**
   * Icon group
   */
  iconGroup: PropTypes.string,
  /**
   * Image url
   */
  image: PropTypes.string,
  /**
   * Offline/local server image filename
   */
  localImg: PropTypes.string,
  /**
   * If true will try to fetch local/offline images
   */
  offlineEnabled: PropTypes.bool,
  /**
   * Thumb width in pixels
   */
  width: PropTypes.number.isRequired,
};

export default Thumb;


styleguide.config.js
const path = require('path');

module.exports = {
  title: '@ox2/thumb',
  serverPort: 55080,
  styleguideDir: './docs',
  skipComponentsWithoutExample: true,
  webpackConfig: {
    module: {
      loaders: [
        {
          test: /\.js?$/,
          exclude: /node_modules/,
          use: {
            loader: 'babel-loader',
            options: {
              presets: ['env'],
              plugins: [
                require('babel-plugin-transform-class-properties'),
                require('babel-plugin-transform-do-expressions'),
                require('babel-plugin-transform-export-extensions'),
                require('babel-plugin-transform-object-rest-spread'),
              ],
            },
          },
        },
        {
          test: /\.css$/,
          use: ['style-loader', 'css-loader'],
        },
      ],
    },
  },
  require: [
    '@ox2/css-font-roboto-condensed',
    '@ox2/css-font-roboto',
    '@ox2/ycss',
  ],
  sections: [
    {
      name: 'Get Started',
      sections: [
        {
          name: 'About',
          content: './info/introduction.md',
        },
        {
          name: 'Installation',
          content: './info/installation.md',
        },
        {
          name: 'Usage',
          content: './info/usage.md',
        },
      ],
    },
    {
      name: 'Components',
      components: 'src/**/[A-Z]*.js',
    },
    {
      name: 'History',
      content: './CHANGELOG.md',
    },
    {
      name: 'License',
      content: './LICENSE',
    },
  ],
  getExampleFilename: componentpath =>
    componentpath.replace(/\.js$/, '.examples.md'),
  getComponentPathLine: componentPath => {
    const name = path.basename(componentPath, '.js');
    const componentPathFixed = componentPath.replace(/src\//i, '');
    const dir = path.dirname(componentPathFixed);
    return `import ${name} from '@ox2/thumb/${dir}';`;
  },
};


src/Thumb/Thumb.story.js
import React from 'react';
import { storiesOf } from '@storybook/react';
import Thumb from './Thumb';

const image1 =
  'https://randomuser.me/api/portraits/women/50.jpg';
const image2 =
  'https://randomuser.me/api/portraits/women/18.jpg';
const logo1 =
  'https://upload.wikimedia.org/wikipedia/commons/thumb/c/c6/Rohde_%26_Schwarz_Logo.svg/800px-Rohde_%26_Schwarz_Logo.svg.png';

const localImg = 'ktLb8Hqxo28WdToq5.jpg';

storiesOf('Thumb', module)
  .add('square image', () =>
    <Thumb height={60} image={image1} width={60} />,
  )
  .add('offline square image', () =>
    <Thumb
      height={60}
      image={image1}
      localImg={localImg}
      offlineEnabled={true}
      width={60}
    />,
  )
  .add('landscape logo', () =>
    <Thumb height={60} image={logo1} width={120} />,
  )
  .add('square white bg image', () =>
    <Thumb height={60} image={image2} width={60} />,
  )
  .add('icon placeholder', () =>
    <Thumb
      height={60}
      icon="user"
      iconGroup="linear"
      width={60}
    />,
  )
  .add('icon placeholder colored', () =>
    <Thumb
      height={60}
      icon="user"
      iconColor={'brand'}
      iconGroup="linear"
      width={60}
    />,
  )
  .add('icon placeholder dark bg', () =>
    <Thumb
      color={'neutral'}
      height={60}
      icon="user"
      iconColor={'white'}
      iconGroup="linear"
      width={60}
    />,
  )
  .add('icon placeholder small', () =>
    <Thumb
      height={30}
      icon="user"
      iconGroup="linear"
      width={30}
    />,
  )
  .add('square image 80', () =>
    <Thumb height={80} image={image1} width={80} />,
  )
  .add('square 80 icon placeholder', () =>
    <Thumb
      height={80}
      icon="user"
      iconGroup="linear"
      width={80}
    />,
  );
